### ISSUE TITLE
Should we use 2D block scaling in weight quantization for better training convergence?

### ISSUE THREAD

**Issue by TheTinyTeddy** [thu-ml/TetraJet-MXFP4Training#2]
As mentioned in the new Nvidia's FP4 paper https://arxiv.org/pdf/2509.25149, where 2D block scaling has improved the FP4 training convergence since the the backpropagation is computed for the same quantized representation of the weight (32x32 so its transpose is unaffected by quantization in backward pass). 

However, my understanding is that 1D weight quantization in forward and the 1D weight re-quantization in the backward pass (as in your paper) should produce better quantization precision since the quantization granularity is much finer (e.g. 1x32 for 1D vs 32x32 for 2D).

So I was wondering your thought on this, and if their interpretation/implementation of 1D weight quantization was flawed? 

Many thanks!

**Comment by cyx0406** [comment]
Thanks for your question! According to our experience, 1D granularity for the weight could yield better performance than 2D blocks. As for the different 1D block shapes in Fwd. and Bwd., the "re-quantization" ("double quantization" in our paper) is proposed to ensure the correction of the optimization goal & unbiasedness of gradients. 

It is possible that Nvidia's FP4 recipe did not consider this method to align the non-square 1D block shape forward and backward. And another possible reason is that this "double-quantization" method may need careful design in real hardware to avoid much overhead, because we need de-quantization before it is re-quantized in the backward. Our current work has not studied the implementation on Blackwell, and we hope this problem could be studied in future work.

**Comment by TheTinyTeddy** [comment]
Thank you for the reply! 

What do you think they have done instead of "double quantization" for 1D weight quantization? What steps they have omitted in order to reduce the computation overhead (such that it deviates from the correct backprop equations derived in your paper)?

**Comment by cyx0406** [comment]
If you adopt 2D quantization rather than 1D, the block would be the same in the fwd./bwd., so there is no need for the "double quantization". Just using the same matrix is OK for the multiplication. I think their work uses this 2D quantization to avoid misalignment of the weight quantization in fwd./bwd., but maybe sacrificing the accuracy.

I actually can only give such a view on their work, since I haven't carefully seen their implementation. I think you may try to get in touch with them to get more information on the detailed technique in NVIDIA's work.

**Comment by TheTinyTeddy** [comment]
> If you adopt 2D quantization rather than 1D, the block would be the same in the fwd./bwd., so there is no need for the "double quantization". Just using the same matrix is OK for the multiplication. I think their work uses this 2D quantization to avoid misalignment of the weight quantization in fwd./bwd., but maybe sacrificing the accuracy.
> 
> I actually can only give such a view on their work, since I haven't carefully seen their implementation. I think you may try to get in touch with them to get more information on the detailed technique in NVIDIA's work.

FYI
"No, we do not use requantization when using 1D blocks. Both the regular and transposed data are produced from the same BF16 version of the weights. Please keep in mind that the recipe proposed in this paper is quite different to what we are doing. First of all they are using MXFP4 datatype as opposed to us using nvFP4 datatype. Those 2 types are very different. MXFP4 uses blocks of size 32 and (what is probably the most important here) the E8M0 scaling factor (a power of 2). nvFP4 uses blocks of size 16 (and so the 2D scheme uses 16x16=256 elements per scaling factor rather than 1024 which would be the case for MXFP4) and the scaling factor is of the type E4M3 (which enables more precise quantization, but also introduces larger difference between tensors when requantizing to get the transpose).
The experiment shown in the "Pretraining Large Language Models with NVFP4" is also looking at the larger model (12B) and token horizon (10T tokens), which exposes more differences between recipes - a setting that may not matter for a small model may impact a larger model final accuracy."

**Comment by cyx0406** [comment]
Thanks for your response! We would like to share with you our new paper TetraJet-v2: https://arxiv.org/abs/2510.27527


In this paper we adopt fully NVFP4 setting for LLMs, and our ablation study in this paper validates the 1D vs 2D problem and the requantization. We wish it would address your concerns.

**Comment by TheTinyTeddy** [comment]
Thank you for another excellent paper! Will the training code be released for use?

**Comment by cyx0406** [comment]
Thanks for your interest! We plan to release the training code around the time of our future conference submission or after acceptance.
